ghcide/src/Development/IDE/Core/PluginUtils.hs

```diff
@@ -0,0 +1,108 @@
+                                                       IdeState (shakeExtras),
+                                                       mkDelayedAction,
+                                                       shakeEnqueue)
+import qualified Development.IDE.Core.Shake           as Shake
+import           Development.IDE.GHC.Orphans          ()
+import           Development.IDE.Graph                hiding (ShakeValue)
+import           Development.IDE.Types.Location       (NormalizedFilePath)
+import qualified Development.IDE.Types.Location       as Location
+import qualified Development.IDE.Types.Logger         as Logger
+import qualified Ide.PluginUtils                      as PluginUtils
+import qualified Language.LSP.Types                   as LSP
+import           Prettyprinter
+import           Prettyprinter.Render.Text            (renderStrict)
+
+-- ----------------------------------------------------------------------------
+-- Plugin Error wrapping
+-- ----------------------------------------------------------------------------
+
+data GhcidePluginError
+    = forall a . Show a => FastRuleNotReady a
+    | forall a . Show a => RuleFailed a
+    | CoreError PluginUtils.PluginError
+
+instance Pretty GhcidePluginError where
+  pretty = \case
+    FastRuleNotReady rule -> "FastRuleNotReady:" <+> viaShow rule
+    RuleFailed rule       -> "RuleFailed:" <+> viaShow rule
+    CoreError perror      -> pretty $ PluginUtils.prettyPluginError perror
+
+pluginResponse ::
+    Monad m =>
+    ExceptT GhcidePluginError m a ->
+    m (Either LSP.ResponseError a)
+pluginResponse = PluginUtils.pluginResponse' handlePluginError
+
+withPluginError :: Functor m => ExceptT PluginUtils.PluginError m a -> ExceptT GhcidePluginError m a
+withPluginError = PluginUtils.withError CoreError
+
+mkPluginErrorMessage :: T.Text -> GhcidePluginError
+mkPluginErrorMessage = CoreError . PluginUtils.mkPluginErrorMessage
+
+handlePluginError :: GhcidePluginError -> LSP.ResponseError
+handlePluginError msg = PluginUtils.mkSimpleResponseError $ renderStrict simpleDoc
+    where
+        simpleDoc = layoutPretty defaultLayoutOptions $ pretty msg
+
+-- ----------------------------------------------------------------------------
+-- Action wrappers
+-- ----------------------------------------------------------------------------
+
+runAction :: MonadIO m => String -> IdeState -> ExceptT e Action a -> ExceptT e m a
+runAction herald ide act =
+  PluginUtils.hoistExceptT . ExceptT $
+    join $ shakeEnqueue (shakeExtras ide) (mkDelayedAction herald Logger.Debug $ runExceptT act)
+
+-- | Request a Rule result, it not available return the last computed result which may be stale.
+--   Errors out if none available.
+useWithStale_ ::(IdeRule k v)
+    => k -> NormalizedFilePath -> ExceptT e Action (v, PositionMapping)
+useWithStale_ key file = ExceptT $ fmap Right $ Shake.useWithStale_ key file
+
+useWithStale :: IdeRule k v
```

Comment:
Okay, so serious question: since we're in `Action` here, and it's safe to throw exceptions, is there any reason not to throw these errors as exceptions too? 

That would give us an (I think) sensible heuristic:
- Functions that run in `Action` throw exceptions
- Functions that run outside `Action` use `ExceptT`

Reply:
That also suggests that the functions here could just become "stricter" versions of `useWithStale` and friends, which throw `RuleNotReady` instead of `RuleFailed` or whatever. Then this module mostly goes away and we instead add a few functions to `Shake.hs`